internal/mode/static/state/graph/route_common.go

```diff
@@ -0,0 +1,567 @@
+package graph
+
+import (
+	"fmt"
+	"strings"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	v1 "sigs.k8s.io/gateway-api/apis/v1"
+	v1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	"github.com/nginxinc/nginx-gateway-fabric/internal/framework/conditions"
+	staticConds "github.com/nginxinc/nginx-gateway-fabric/internal/mode/static/state/conditions"
+	"github.com/nginxinc/nginx-gateway-fabric/internal/mode/static/state/validation"
+)
+
+const wildcardHostname = "~^"
+
+// ParentRef describes a reference to a parent in a Route.
+type ParentRef struct {
+	// Attachment is the attachment status of the ParentRef. It could be nil. In that case, NGF didn't attempt to
+	// attach because of problems with the Route.
+	Attachment *ParentRefAttachmentStatus
+	// Gateway is the NamespacedName of the referenced Gateway
+	Gateway types.NamespacedName
+	// Idx is the index of the corresponding ParentReference in the Route.
+	Idx int
+}
+
+// ParentRefAttachmentStatus describes the attachment status of a ParentRef.
+type ParentRefAttachmentStatus struct {
+	// AcceptedHostnames is an intersection between the hostnames supported by an attached Listener
+	// and the hostnames from this Route. Key is listener name, value is list of hostnames.
+	AcceptedHostnames map[string][]string
+	// FailedCondition is the condition that describes why the ParentRef is not attached to the Gateway. It is set
+	// when Attached is false.
+	FailedCondition conditions.Condition
+	// Attached indicates if the ParentRef is attached to the Gateway.
+	Attached bool
+}
+
+type RouteType string
+
+const (
+	// RouteTypeHTTP indicates that the RouteType of the L7Route is HTTP
+	RouteTypeHTTP RouteType = "http"
+	// RouteTypeGRPC indicates that the RouteType of the L7Route is gRPC
+	RouteTypeGRPC RouteType = "grpc"
+)
+
+// RouteKey is the unique identifier for a L7Route
+type RouteKey struct {
+	NamespacedName types.NamespacedName
+	RouteType      RouteType
+}
+
+// L7Route is the generic type for the layer 7 routes, HTTPRoute and GRPCRoute
+type L7Route struct {
+	// Source is the source Gateway API object of the Route.
+	Source client.Object
+	// RouteType is the type (http or grpc) of the Route.
+	RouteType RouteType
+	// Spec is the L7RouteSpec of the Route
+	Spec L7RouteSpec
+	// ParentRefs describe the references to the parents in a Route.
+	ParentRefs []ParentRef
+	// SrcParentRefs contain the Gateway API references to the parents in a Route.
+	SrcParentRefs []v1.ParentReference
```

Comment:
Why do we need both ParentRefs and SrcParentRefs?

Reply:
Yes we need both, we use the SrcParentRefs in `bindRouteToListeners` and in `prepareRouteStatus`

> would it make sense to move src ParentRefs to the Spec?

Yes, that does make sense. I'll move it 👍🏼